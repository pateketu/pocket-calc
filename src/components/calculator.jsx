import React from 'react';
import Display from './display';
import KeyPad from './key-pad';

export default function Calculator() {
	return	<div className="calculator">
		<div className="ee-branding"></div>
		<Display></Display>
		<KeyPad></KeyPad>
	</div>;
}