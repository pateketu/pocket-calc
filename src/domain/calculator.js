export const Operation = {
	UNKNNOWN:'UNKNOWN',
	DIVIDE:'รท',
	MULTIPLY:'X',
	SUBTRACT:'-',
	ADD:'+',
	EQUAL:'=',
	PERCENTAGE:'%',
	POSITIVE_NEGATIVE:'+/-',
};

const calcOperation = {
	[Operation.DIVIDE]: (currentVal, nextVal) =>  {				
		if(currentVal === 0)
			throw new Error('Error: Cannot divide by Zero');

		
		return !isNaN(parseInt(currentVal)) && !isNaN(parseInt(nextVal)) 
			?  nextVal/currentVal : 
			currentVal || nextVal;
	},
	[Operation.MULTIPLY]: (currentVal, nextVal) => isNaN(parseInt(nextVal)) ? currentVal : currentVal * nextVal,
	[Operation.SUBTRACT]:(currentVal, nextVal) =>
		!isNaN(parseInt(currentVal)) && !isNaN(parseInt(nextVal)) ?  nextVal - currentVal
			: currentVal || nextVal,
	[Operation.ADD]:(currentVal, nextVal) =>isNaN(parseInt(nextVal)) ? currentVal : currentVal + nextVal,
	[Operation.EQUAL]:(currentVal) => currentVal,
	[Operation.PERCENTAGE]:(currentVal) => currentVal/100,
	[Operation.POSITIVE_NEGATIVE]:(currentVal) => currentVal < 0 ? Math.abs(currentVal) : currentVal * -1
};

export function calculate(operation, currentVal, nextVal){
	if(!(operation in calcOperation)) return;

	if(isNaN(parseInt(currentVal))) return;

	if(nextVal !== undefined && isNaN(parseInt(nextVal))) return;

	return fixFloatPrecision(calcOperation[operation](currentVal, nextVal));
}

function fixFloatPrecision(result){
	let numericVal = result;	
	if( Math.abs(numericVal) % 1 !== 0){
		numericVal = numericVal.toString().split('.')[1].length >=2 
			? numericVal.toFixed(2) : numericVal;
	}	
	return Number(numericVal);
}
