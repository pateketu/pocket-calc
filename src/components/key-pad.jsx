import React from 'react';
import {Operation} from '../domain/calculator';
import Key from './key';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import * as actions from '../redux/calcActions';
export class KeyPad extends React.PureComponent{
     
	constructor(props){
		super(props);
		this.createOperationKeys();
		this.createNumberKeys();
		this.createArthmaticOperationKeys();
	}

	createOperationKeys(){
		this.operationKeys = {
			'C': () => this.props.clear(),
			[Operation.POSITIVE_NEGATIVE] : () => this.props.operation(Operation.POSITIVE_NEGATIVE),
			[Operation.PERCENTAGE]: () => this.props.operation(Operation.PERCENTAGE)
		};
	}

	createNumberKeys(){
		this.numberKeys = [
			{key:0, action:() => this.props.input(0)},
			{key:'.', action:() => this.props.point()}
		];	
		
		for(let i=1;i<=9;i++){
			this.numberKeys.push({key:i, action:()=>this.props.input(i)});
		}
	}

	createArthmaticOperationKeys(){
		this.arthmaticOperationKeys = {
			[Operation.DIVIDE]: ()=>this.props.operation(Operation.DIVIDE),
			[Operation.MULTIPLY]: ()=>this.props.operation(Operation.MULTIPLY),
			[Operation.SUBTRACT]: ()=>this.props.operation(Operation.SUBTRACT),
			[Operation.ADD]: ()=>this.props.operation(Operation.ADD),
			[Operation.EQUAL]: ()=>this.props.operation(Operation.EQUAL)
		};
	}

	handleWithCatch = (func)=>{
		try{
			func();
		}catch(error){
			this.props.error(error);
		}
	}
	
	render(){
		return <div className="d-flex keyPad">
			<div className="keyPad-col-1">
				<div className="d-flex operationKeys">
					{ 
						Object.keys(this.operationKeys).map(key =>
							<Key key={key} 
								value={key} 
								onClick={()=> this.handleWithCatch(()=> 
									this.operationKeys[key]())}>
							</Key>)
					}
				</div>
				<div className="d-flex flex-wrap-reverse numberKeys">
					{
						this.numberKeys.map(numberKey=>
							<Key key={numberKey.key} 
								className={`number-key-${numberKey.key}`}
								value={numberKey.key.toString()} 
								onClick={()=> numberKey.action()}>
							</Key>)
					}
				</div>
			</div>
			<div className="d-flex flex-column arthmaticOperationKeys">
				{
					Object.keys(this.arthmaticOperationKeys).map(key =>
						<Key key={key} 
							value={key} 
							isActive = {this.props.currentOperation === key}
							onClick={()=> this.handleWithCatch(()=> 
								this.arthmaticOperationKeys[key]())}>
						</Key>)
				}
			</div>
		</div>;
	}    
}

export default connect((state)=>({
	currentOperation:state.currentOperation
}),
(dispath)=>({
	operation:(ops)=>dispath(actions.operation(ops)),
	clear:()=>dispath(actions.clear()),
	input:(number)=>dispath(actions.input(number)),
	point:()=>dispath(actions.point()),
	error:(error)=>dispath(actions.error(error))
}))(KeyPad);

KeyPad.propTypes = {
	currentOperation:PropTypes.string,
	clear:PropTypes.func.isRequired,
	operation:PropTypes.func.isRequired,
	input:PropTypes.func.isRequired,
	point:PropTypes.func.isRequired,
	error:PropTypes.func.isRequired
};
