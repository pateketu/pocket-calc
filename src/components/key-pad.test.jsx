import * as React from 'react';
import {KeyPad} from './key-pad';
import {mount} from 'enzyme';
import { Operation } from '../domain/calculator';
import toJson from 'enzyme-to-json';
describe('key pad component', ()=>{
	
	it('renders as expected', ()=>{
		const wrapper = mount(<KeyPad 
			operation={()=>{}} 
			clear={()=>{}}
			input={()=>{}}
			point={()=>{}}
			error={()=>{}}></KeyPad>);	
		expect(toJson(wrapper, {noKey:true})).toMatchSnapshot();   
		
	});
	describe('actions', ()=>{
		it('pressing clear button, dispatches clear action', ()=>{			
			const mockClear =  jest.fn();
			const wrapper = mount(<KeyPad 
				operation={()=>{}} 
				input={()=>{}}
				point={()=>{}}
				clear={mockClear}
				error={()=>{}}></KeyPad>);				
			wrapper.find('button[children="C"]').simulate('click');
			expect(mockClear).toHaveBeenCalled();
		});

		it('pressing operation button, dispatches expected operation action', ()=>{					
			
			[Operation.POSITIVE_NEGATIVE, Operation.PERCENTAGE].forEach((operation)=>{
				const mockOperation =  jest.fn();
				const wrapper = mount(<KeyPad 
					operation={mockOperation} 
					clear={()=>{}}
					input={()=>{}}
					point={()=>{}}
					error={()=>{}}></KeyPad>);				
				wrapper.find(`button[children="${operation}"]`).simulate('click');
				expect(mockOperation).toHaveBeenCalledWith(operation);
			});
			
		});

		it('pressing number button, dispatches expected input action', ()=>{		
			
			for(let i=0;i<=9;i++){
				const mockInput =  jest.fn();
				const wrapper = mount(<KeyPad 
					operation={()=>{}} 
					clear={()=>{}}
					input={mockInput}
					point={()=>{}}
					error={()=>{}}></KeyPad>);								
				wrapper.find(`button[children="${i}"]`).simulate('click');
				expect(mockInput).toHaveBeenCalledWith(i);
			}
		});
		it('shows currentOperation as active', ()=>{

			const currentOpertaion =Operation.ADD; 
			
			const wrapper = mount(<KeyPad 
				currentOperation = {currentOpertaion}
				operation={()=>{}} 
				clear={()=>{}}
				input={()=>{}}
				point={()=>{}}
				error={()=>{}}
			></KeyPad>);							
			
			wrapper.find(`button[children="${currentOpertaion}"]`).hasClass('active');
		});
	});
	
	
});