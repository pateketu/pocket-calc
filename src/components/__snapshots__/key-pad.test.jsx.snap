// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`key pad component renders as expected 1`] = `
<KeyPad
  clear={[Function]}
  error={[Function]}
  input={[Function]}
  operation={[Function]}
  point={[Function]}
>
  <div
    className="d-flex"
  >
    <div>
      <div
        className="d-flex"
      >
        <Key
          onClick={[Function]}
          value="C"
        >
          <button
            className="btn btn-outline-secondary btn-lg"
            onClick={[Function]}
          >
            C
          </button>
        </Key>
        <Key
          onClick={[Function]}
          value="+/-"
        >
          <button
            className="btn btn-outline-secondary btn-lg"
            onClick={[Function]}
          >
            +/-
          </button>
        </Key>
        <Key
          onClick={[Function]}
          value="%"
        >
          <button
            className="btn btn-outline-secondary btn-lg"
            onClick={[Function]}
          >
            %
          </button>
        </Key>
      </div>
      <div
        className="d-flex flex-wrap-reverse"
      >
        <Key
          className="number-key-0"
          onClick={[Function]}
          value="0"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-0"
            onClick={[Function]}
          >
            0
          </button>
        </Key>
        <Key
          className="number-key-."
          onClick={[Function]}
          value="."
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-."
            onClick={[Function]}
          >
            .
          </button>
        </Key>
        <Key
          className="number-key-1"
          onClick={[Function]}
          value="1"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-1"
            onClick={[Function]}
          >
            1
          </button>
        </Key>
        <Key
          className="number-key-2"
          onClick={[Function]}
          value="2"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-2"
            onClick={[Function]}
          >
            2
          </button>
        </Key>
        <Key
          className="number-key-3"
          onClick={[Function]}
          value="3"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-3"
            onClick={[Function]}
          >
            3
          </button>
        </Key>
        <Key
          className="number-key-4"
          onClick={[Function]}
          value="4"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-4"
            onClick={[Function]}
          >
            4
          </button>
        </Key>
        <Key
          className="number-key-5"
          onClick={[Function]}
          value="5"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-5"
            onClick={[Function]}
          >
            5
          </button>
        </Key>
        <Key
          className="number-key-6"
          onClick={[Function]}
          value="6"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-6"
            onClick={[Function]}
          >
            6
          </button>
        </Key>
        <Key
          className="number-key-7"
          onClick={[Function]}
          value="7"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-7"
            onClick={[Function]}
          >
            7
          </button>
        </Key>
        <Key
          className="number-key-8"
          onClick={[Function]}
          value="8"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-8"
            onClick={[Function]}
          >
            8
          </button>
        </Key>
        <Key
          className="number-key-9"
          onClick={[Function]}
          value="9"
        >
          <button
            className="btn btn-outline-secondary btn-lg number-key-9"
            onClick={[Function]}
          >
            9
          </button>
        </Key>
      </div>
    </div>
    <div
      className="d-flex flex-column"
    >
      <Key
        isActive={false}
        onClick={[Function]}
        value="รท"
      >
        <button
          className="btn btn-outline-secondary btn-lg"
          onClick={[Function]}
        >
          รท
        </button>
      </Key>
      <Key
        isActive={false}
        onClick={[Function]}
        value="X"
      >
        <button
          className="btn btn-outline-secondary btn-lg"
          onClick={[Function]}
        >
          X
        </button>
      </Key>
      <Key
        isActive={false}
        onClick={[Function]}
        value="-"
      >
        <button
          className="btn btn-outline-secondary btn-lg"
          onClick={[Function]}
        >
          -
        </button>
      </Key>
      <Key
        isActive={false}
        onClick={[Function]}
        value="+"
      >
        <button
          className="btn btn-outline-secondary btn-lg"
          onClick={[Function]}
        >
          +
        </button>
      </Key>
      <Key
        isActive={false}
        onClick={[Function]}
        value="="
      >
        <button
          className="btn btn-outline-secondary btn-lg"
          onClick={[Function]}
        >
          =
        </button>
      </Key>
    </div>
  </div>
</KeyPad>
`;
